File: Depth-First Search/0270.Closest Binary Search Tree Value/Solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def closestValue(self, root: Optional[TreeNode], target: float) -> int:
        ans = float('inf')

        def dfs(root):
            nonlocal ans, target
            if not root:
                return
            if abs(root.val - target) <= abs(ans - target):
                ans = root.val
            if root.val < target:
                dfs(root.right)
            else:
                dfs(root.left)

        dfs(root)
        return ans

File: Depth-First Search/0270.Closest Binary Search Tree Value/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int ans = Integer.MAX_VALUE;
    double target;

    public int closestValue(TreeNode root, double target) {
        this.target = target;
        dfs(root);
        return ans;
    }

    public void dfs(TreeNode root) {
        if (root == null) {
            return;
        }
        if (Math.abs(root.val - target) < Math.abs(ans - target)) {
            ans = root.val;
        }
        if (root.val < target) {
            dfs(root.right);
        } else {
            dfs(root.left);
        }
    }
}